src/graph_stats.cc

```diff
@@ -0,0 +1,160 @@
+#include "graph_stats.h"
+#include "state.h"
+#include "graph.h"
+
+// Standard headers
+#include <limits>
+#include <set>
+#include <queue>
+#include <vector>
+#include <cassert>
+
+namespace {
+
+class GetGraphHeight
+{
+public:
+  GetGraphHeight()
+    : _heights()
+  {}
+
+private:
+  typedef map<Node*, size_t> Heights;
+  Heights _heights;
+
+public:
+  void Reset() {
+    _heights.clear();
+  }
+
+  size_t operator()(const vector<Node*>& root_nodes) {
+    Reset();
+    return GetMaxHeight(root_nodes);
+  }
+
+private:
+  size_t operator()(Node* node) {
+    size_t& height = _heights[node];
+    if (height == 0) {
+      if (Edge* e = node->in_edge()) {
+        height = GetMaxHeight(e->inputs_) + 1;
+      } else {
+        height = 1;
+      }
+    }
+    return height;
+  }
+
+  size_t GetMaxHeight(const vector<Node*>& nodes) {
+    size_t height = numeric_limits<size_t>::min();
+    for (vector<Node*>::const_iterator i = nodes.begin();
+         i != nodes.end(); ++i) {
+      height = max(height, (*this)(*i));
+    }
+    return height;
+  }
+
+}; // class GetGraphHeight
+
+class GetGraphWidth
+{
+public:
+  GetGraphWidth()
+    : _q()
+    , _s()
+  {}
+
+  void Reset() {
+    _s.clear();
+  }
+
+  size_t operator()(const vector<Node*>& root_nodes) {
+    Reset();
+    assert(_q.empty());
+    _PushNode(root_nodes);
+    size_t w = _q.size();
+    while (!_q.empty()) {
+      Node* n = _q.front();
+      _q.pop();
+      if (Edge* e = n->in_edge()) {
+        _PushNode(e->inputs_);
+        w = max(w, _q.size());
+      }
+    }
+    return w;
+  }
+
+private:
+  void _PushNode(const vector<Node*>& nodes) {
+    for (vector<Node*>::const_iterator i = nodes.begin();
+         i != nodes.end(); ++i) {
+      pair<set<Node*>::iterator, bool> ret = _s.insert(*i);
+      if (ret.second)
+        _q.push(*i);
+    }
+  }
+
+  queue<Node*> _q;
+  set<Node*> _s;
+}; // class GetGraphWidth
+
+} // anonymous namespace
+
+void GraphStats::Reset() {
```

Comment:
If the default constructor and this function do the same thing, wouldn't be better to not have this function at all and just create a new object when you want the values reset?